### PR TITLE
Utilize standalone rimage tool

### PR DESCRIPTION
**Pull request by dcpleung**
This modifies the build script to utilize an externally built `rimage` tool, and also removes the `rimage` tool from source tree.

Requires https://github.com/thesofproject/rimage/pull/1 and https://github.com/thesofproject/rimage/pull/3

### CODE REVIEW

**Comment by plbossart** [comment]
Why is this needed?

**Comment by jajanusz** [comment]
@plbossart We want to decouple rimage from SOF as it is not audio-specific stuff, it's just image composing tool for DSP's FW. It's weird f.e. for Zephyr to pull whole SOF repo just for rimage.

**Comment by lgirdwood** [comment]
@jajanusz @dcpleung it make sense to create a subrepo for rimage in sof. This keep everything in sync with th eexternal repo.

**Comment by jajanusz** [comment]
> @jajanusz @dcpleung it makes sense to create a subrepo for rimage in sof. This keep everything in sync with the external repo.

Yea good idea, git submodule will make things easier

**Comment by dcpleung** [comment]
Understood. I will work on getting rimage repo as a submodule.

**Comment by plbossart** [comment]
Are you really sure git submodules are the answer?
I have seen nothing but issues with submodules, and I would like to double-check this is compatible with GitHub (PRs, merge, etc)

see e.g. https://github.blog/2016-02-01-working-with-submodules/ 

If we go with multiple repositories them it's just easier to use "repo".

**Comment by dcpleung** [comment]
Updated to use git submodule.

**Comment by lgirdwood** [comment]
@plbossart submodules do have a downside in that users need to remeber and manually update each submodule when updating the main repo. A script would eliminate any forgetfullnes though (which qemu do run for their 3 submodules as part of the build process so maybe this is the best method). 
Fwiw, repo has a manual cost too since we need to keep the manifest updated.

**Comment by lgirdwood** [comment]
@xiulipan @zrombel do we need any CI changes to deal with a rimage git submodule or would including submodule update in part of build be good enough for CI.

**Comment by dcpleung** [comment]
> @dcpleung could rimage be updated via a script and could this script be invoked at the cmake stage where we set toolchain etc. I'm trying to find a way that wont impact CI and means users can test local rimage updates.
> 
> ```
> mkdir build_byt && cd build_byt
> cmake -DTOOLCHAIN=xtensa-byt-elf -DROOT_DIR=`pwd`/../../xtensa-root/xtensa-byt-elf ..
> ```. 
> ```

It is possible to add a cmake custom command/target to run git submodule update or the script. However, I am not sure the impact since I haven't used submodule much.

**Comment by plbossart** [comment]
> @plbossart submodules do have a downside in that users need to remeber and manually update each submodule when updating the main repo. A script would eliminate any forgetfullnes though (which qemu do run for their 3 submodules as part of the build process so maybe this is the best method).

Indeed, we need to take the user out of the equation. It's got to be scripted/automated.

> Fwiw, repo has a manual cost too since we need to keep the manifest updated.

not really, and we already have one manifest btw https://github.com/thesofproject/sof-manifests

**Comment by jajanusz** [comment]
> > @dcpleung could rimage be updated via a script and could this script be invoked at the cmake stage where we set toolchain etc. I'm trying to find a way that wont impact CI and means users can test local rimage updates.
> > ```
> > mkdir build_byt && cd build_byt
> > cmake -DTOOLCHAIN=xtensa-byt-elf -DROOT_DIR=`pwd`/../../xtensa-root/xtensa-byt-elf ..
> > ```. 
> > ```
> 
> It is possible to add a cmake custom command/target to run git submodule update or the script. However, I am not sure the impact since I haven't used submodule much.

You basically can do it similar to what we do in test/cmocka, just for download command of ExternalProject instead of git clone use submodule init/update.

**Comment by xiulipan** [comment]
@lgirdwood We will use scripts to do the build.
Please also update the build scripts after we have make an alignment about how to deal with the rimage sync with sof.

**Comment by jajanusz** [comment]
@xiulipan @dcpleung @lgirdwood @plbossart You need no changes in CI, everything can be done in cmake files if done properly

**Comment by lgirdwood** [comment]
@jajanusz I was hoping you would chime in here. :) Ideally we want to synchronise any external submodules prior to a build but also need to be able to disable this and force a local rimage verion (for say anyone hacking and testing rimage). How would we do this in cmake ?

**Comment by jajanusz** [comment]
> @jajanusz I was hoping you would chime in here. :) Ideally we want to synchronise any external submodules prior to a build but also need to be able to disable this and force a local rimage verion (for say anyone hacking and testing rimage). How would we do this in cmake ?
> 
> To give an example, qemu does submodule updates as part of "configure" which is a python script in qemu.
> 
> ```
> ./configure (updates to latest submodules)
> make
> make install
> ```
> 
> configure is only needed to be run when changing the qemu configuration or if you need to update it's submodules. It also allows for pulling certain commits/tags of submodules.
> 
> I guess we could have this as part of a make rule or as part of the logic for generating the makefiles ?

@dcpleung @lgirdwood You need mix of:
https://github.com/thesofproject/sof/blob/decd2064607d74106dc176bfd04a4f4a4e584eba/test/cmocka/CMakeLists.txt#L14-L28
&
https://github.com/thesofproject/sof/blob/decd2064607d74106dc176bfd04a4f4a4e584eba/src/arch/xtensa/CMakeLists.txt#L205-L215
in the place of current rimage_ep target, it should be easy, just fix paths and adujst arguments (I assume rimage in other repo is also cmake proj).
If you want I can do it in beginning of december, cos I'm going for 2 weeks holiday tomorrow ;p

**Comment by jajanusz** [comment]
@dcpleung @lgirdwood I guess it'd be something like this but I didn't check it.
```cmake
ExternalProject_Add(rimage_ep 
 	DEPENDS check_version_h 
 	GIT_REPOSITORY https://github.com/thesofproject/rimage # ??
 	PREFIX "${PROJECT_BINARY_DIR}/rimage_ep" 
 	BINARY_DIR "${PROJECT_BINARY_DIR}/rimage_ep/build" 
 	BUILD_ALWAYS 1 
 	CMAKE_ARGS -DVERSION_H_PATH=${VERSION_H_PATH} 
 		-DPEM_KEY_PREFIX=${PROJECT_SOURCE_DIR}/rimage/keys 
 	INSTALL_COMMAND "" 
 ) 
```

**Comment by jajanusz** [comment]
> @dcpleung @lgirdwood I guess it'd be something like this but I didn't check it.
> 
> ```cmake
> ExternalProject_Add(rimage_ep 
>  	DEPENDS check_version_h 
>  	GIT_REPOSITORY https://github.com/thesofproject/rimage # ??
>  	PREFIX "${PROJECT_BINARY_DIR}/rimage_ep" 
>  	BINARY_DIR "${PROJECT_BINARY_DIR}/rimage_ep/build" 
>  	BUILD_ALWAYS 1 
>  	CMAKE_ARGS -DVERSION_H_PATH=${VERSION_H_PATH} 
>  		-DPEM_KEY_PREFIX=${PROJECT_SOURCE_DIR}/rimage/keys 
>  	INSTALL_COMMAND "" 
>  ) 
> ```

Ofc it's for pure git repo download variant, if you want to go with submodules then instead of `GIT_REPOSTITORY` use something like `DOWNLOAD_COMMAND " git submodule update --init"`

**Comment by dcpleung** [comment]
I can add the `submodule update` if we are still going to the submodule route.

**Comment by lgirdwood** [comment]
> I can add the `submodule update` if we are still going to the submodule route.

Please do.

**Comment by dcpleung** [comment]
Added the download command to do `submodule update`.

**Comment by paulstelian97** [comment]
> @dcpleung, looks good. Which build rule updates and builds rimage (I'm guessing default) ? Also, how would I force it to use my local rimage (if I was hacking rimage) ?

I think "submodule update" will not remove unstaged files in submodules but fail if there is a conflict or make you resolve a merge conflict if there is any local change. Not 100% sure though, but I don't think local change will be automatically discarded.

**Comment by dcpleung** [comment]
> > @dcpleung, looks good. Which build rule updates and builds rimage (I'm guessing default) ? Also, how would I force it to use my local rimage (if I was hacking rimage) ?
> 
> I think "submodule update" will not remove unstaged files in submodules but fail if there is a conflict or make you resolve a merge conflict if there is any local change. Not 100% sure though, but I don't think local change will be automatically discarded.

It is being built by default, so there is no change to the build instructions.

As for changes, you can make changes directly under `rimage/` and also create commits. I tried it out just now, and `submodule update` did nothing to the `rimage/`.

**Comment by lgirdwood** [comment]
SOFCI TEST

**Comment by lgirdwood** [comment]
Running CI again now that new repo has code pushed.

**Comment by lgirdwood** [comment]
@zrombel does internal CI need updated to deal with rimage submodule ?

**Comment by zrombel** [comment]
@lgirdwood quick answer is: yes. Code is cloned on master machine and copied to build machines where build process is being performed and from what I can see in build log build machines doesn't have access to public github server. I will ask what we can do about it and give you feedback.

**Comment by zrombel** [comment]
Build machines cannot have access to public github repo - security issue, but submodules can be updated right after cloning repo. Can CMake skip submodule update if it's already present ?

**Comment by lgirdwood** [comment]
@zrombel not sure about cmake, that's a @jajanusz question. But could your master machine also clone rimage repo too and copy it over to test machine as part of FW copy script ?

**Comment by zrombel** [comment]
Of course, done it.

**Comment by lgirdwood** [comment]
SOFCI TEST

**Comment by lgirdwood** [comment]
@zrombel can you check the build, it failed for all so maybe an issue copying to master build server ?

**Comment by jajanusz** [comment]
Any update? I'd like not to cut out rimage from this repo with how https://github.com/thesofproject/rimage looks now.

**Comment by lgirdwood** [comment]
@jajanusz I'm assuming you mean the need for cmake to build  rimage on Windows ?

**Comment by jajanusz** [comment]
> @jajanusz I'm assuming you mean the need for cmake to build rimage on Windows ?

I just don't want to go back to automake ;)
Anyway what's the plan IRT this PR? @dcpleung

**Comment by lgirdwood** [comment]
I'm happy either way, I do prefer autotools since they are pretty much standard on all Linux, but using cmake does mean it's easier for most folks in the dev team.

**Comment by jajanusz** [comment]
> I'm happy either way, I do prefer autotools since they are pretty much standard on all Linux, but using cmake does mean it's easier for most folks in the dev team.

Both sof and zephyr use cmake already, so it's better to stick to one tool.

**Comment by dcpleung** [comment]
I will update this to use cmake instead of autotools.

**Comment by dcpleung** [comment]
I have pushed a version to use CMake to build the standalone rimage.

Requires https://github.com/thesofproject/rimage/pull/3

File: LICENCE [hunk 1]
```diff
@@ -75,46 +75,3 @@ scripts/kconfig/olddefconfig.py
 # AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 # OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 
-
-Files with 2-Clause BSD licence:
-
-rimage/elf.h
```
Comment (jajanusz) [inline LICENCE]: Why isn't this license copied to rimage repo?

Comment (dcpleung) [reply LICENCE]: There is already a SPDX line in the header. Is that not enough?

Comment (jajanusz) [reply LICENCE]: IANAL, but SPDX is just for source file. LICENSE is for binary distribution and should contains full texts of licenses.
@lgirdwood please comment

Comment (lgirdwood) [reply LICENCE]: Probably safest to include the LICENCE file here.I

Comment (dcpleung) [reply LICENCE]: A PR is created to update the LICENSE file in the rimage repo: https://github.com/thesofproject/rimage/pull/4

**Comment by lgirdwood** [comment]
@mmaka1 I think we just need the manifest headers right ? They could be moved into rimage and SOF can then include them.

**Comment by mmaka1** [comment]
@lgirdwood Actually, a quick and pretty good way to make the rImage really independent (instead of implementing framework for extra steps inside rImage) would be to separate ldc related code (the `HAS_FILE_FORMAT_H` part) to another sof-ldc tool, keep it inside the sof repo and run as extra step in sof makefiles with sof elf as input and ldc as output.

**Comment by lgirdwood** [comment]
@mmaka1 sorry, not following - do you mean a new repo for ldc generation tool ? or include ldc tool with rimage ?

**Comment by mmaka1** [comment]
@lgirdwood The idea was to keep the SOF specific code (ldc file creation part) in the sof repo, factored as a standalone tool run as additional fw image build step. And move only generic part of rImage to the new location. When ldc creation code is isolated, there is no need to "import" anything (header includes) from sof repo into rimage repo neither maintain sof specific code within the rimage source, otherwise there is unnecessary cost of version control, compatibility checking etc. introduced.

**Comment by lgirdwood** [comment]
@mmaka1 @ktrzcinx do you guys have any update on the ldc de-couplement from rimage ?

**Comment by ktrzcinx** [comment]
@lgirdwood I'm currently working on it, there should be PR for this change soon

**Comment by lgirdwood** [comment]
@dcpleung we are good to merge once conflicts are fixed, the LDC part has been removed.

**Comment by dcpleung** [comment]
> @dcpleung we are good to merge once conflicts are fixed, the LDC part has been removed.

I will update the rimage repo then update this PR.

**Comment by dcpleung** [comment]
Updated rimage repo, and this PR. Now requires https://github.com/thesofproject/rimage/pull/5

**Comment by lgirdwood** [comment]
@dcpleung https://github.com/thesofproject/rimage/pull/5 now merged, will run CI again.

**Comment by lgirdwood** [comment]
SOFCI TEST

**Comment by lgirdwood** [comment]
@xiulipan I think we need a simple ```scripts/sof-submodule-update.sh``` that will checkout a version of each submodule. i.e. something like
```
SUBMODULE_RIMAGE_HASH= 123abcdef
...
cd submodule
git checkout SUBMODULE_RIMAGE_HASH
```
This would then allow rimage updates to be versioned and updated with FW/CI.

**Comment by dcpleung** [comment]
I have updated the commit SHA for the submodule.

For CI, if it has a newer git version, you can do "git clone --recurse-submodules" to also clone the submodules at the same time.

**Comment by lgirdwood** [comment]
@xiulipan @zrombel can you guys make the update to CI to get the submodule. I will merge this tiday.

**Comment by dcpleung** [comment]
SOFCI TEST

**Comment by xiulipan** [comment]
@dcpleung @lgirdwood Do we need any other cmd more than `scripts/build-xtensa-all.sh`? I think no update is need for CI as CI is always using the scripts, if scripts is updated, no update for CI is needed.

**Comment by jajanusz** [comment]
> @dcpleung @lgirdwood Do we need any other cmd more than `scripts/build-xtensa-all.sh`? I think no update is need for CI as CI is always using the scripts, if scripts is updated, no update for CI is needed.

It's handled by cmake's external project.
I don't want developers to think about rimage.
They should just write 3 commands `cmake && make *_defconfig && make bin`, if it stays true then you shouldn't worry about CI.

**Comment by xiulipan** [comment]
@jajanusz I see that this is work on bare new repos.
@dcpleung I found some issue with updated folder, reuse the rimage folder seems not a very good idea for old repos. I will try to add some clean up for CI scripts.

**Comment by jajanusz** [comment]
> @jajanusz I see that this is work on bare new repos.
> @dcpleung I found some issue with updated folder, reuse the rimage folder seems not a very good idea for old repos. I will try to add some clean up for CI scripts.

IMHO you should remove `build*` folders to make clean build, otherwise CI may have weird behaviour when 1 run of CI affects another, I ran into these issues in the past already.

**Comment by lgirdwood** [comment]
@xiulipan Jenkins/Travis now good to merge.
@zrombel does Internal CI need anything here to use the git submodule rimage instead of a local rimage ? I notice it's failing FW load tests for IPC major version (which seems unrelated ?)
```
2020-04-06 19:25:59ITERATION TestLoadFwExtended.test_load_fw_and_check_version === 1586193959.7006292
2020-04-06 19:25:59=> TestLoadFwExtended.test_load_fw_and_check_version
2020-04-06 19:25:59     => Loading base firmware
2020-04-06 19:26:00     => Exception in test.
2020-04-06 19:26:00                 Exception AssertionError occurred at 'C:\\work\\tools\\pyexecnetcache\\cavs_scripts-py\\tests\\sof\\fw_00_basic\\test_01_load_fw_extended.py':66
                                    Major version from IPC message does not match major version from firmware header.
                                    assert 1 == 0
                                     +  where 1 = <utilities.fw_structures.FirmwareVersion object at 0x000001766580EA48>.major
                                     +    where <utilities.fw_structures.FirmwareVersion object at 0x000001766580EA48> = <utilities.test_controller.TestController object at 0x00000176656FD4A8>.fw_version
                                     +  and   0 = <utilities.fw_manifest.AdspFwBinaryHeader object at 0x0000017665AEB648>.major_version
                                     +    where <utilities.fw_manifest.AdspFwBinaryHeader object at 0x0000017665AEB648> = <utilities.fw_manager.FwManager object at 0x00000176656749B0>.binary_header
```

**Comment by ktrzcinx** [comment]
@dcpleung please remove conflicts

**Comment by lgirdwood** [comment]
So I guess there are now no blockers on CI side to merge, let merge as soon as conflict are resolved.

**Comment by dcpleung** [comment]
Conflicts resolved and updated to latest commits from rimage.

**Comment by zrombel** [comment]
CI is all good but it seems that there is a problem with building TGL with standalone rimage. I've asked @jajanusz to look into it.

**Comment by jajanusz** [comment]
CI is failing cos rimage is incorrectly parsing ELF, you are missing elf.c changes related to LMA/VMA from this PR (rimage* commits) : https://github.com/thesofproject/sof/pull/2516

**Comment by ktrzcinx** [comment]
@dcpleung I suppose we should add `--remote` flag to `submodule update --init --recursive` command to be aligned with remote repository for each submodule

**Comment by dcpleung** [comment]
Rebased and updated the rimage submodule SHA.

**Comment by ktrzcinx** [comment]
@dcpleung thanks

**Comment by jajanusz** [comment]
@dcpleung 
Great, one nitpick left.
Some commits titles don't follow format used in this repo and in kernel -> `topic: what changed`.
Please update them.
Thanks.

**Comment by marc-hb** [comment]
> Code is cloned on master machine and copied to build machines where build process is being performed and from what I can see in build log build machines doesn't have access to public github server. 

@zrombel Is there an exception for cmocka?
```cmake
        ExternalProject_Add(cmocka_git
                GIT_REPOSITORY https://github.com/thesofproject/cmocka
 ```

**Comment by marc-hb** [comment]
Reworked in `cmake/git-submodules: don't throw away work in progress at build time` #3048

**Comment by zrombel** [comment]
@marc-hb No, CI is using prebuilt cmoka on VM's.
```
if(DEFINED CMOCKA_DIRECTORY)
	# Prebuilt Cmocka lib
	set_property(TARGET cmocka PROPERTY IMPORTED_LOCATION "${CMOCKA_DIRECTORY}/lib/libcmocka-static.a")
	set(CMOCKA_INCLUDE_DIR "${CMOCKA_DIRECTORY}/include")
else()
	set(cmocka_binary_directory "${PROJECT_BINARY_DIR}/cmocka_git/build")

	# Build Cmocka locally
	ExternalProject_Add(cmocka_git
```

**Comment by marc-hb** [comment]
While testing "tarbomb" fix #3132 I noticed the conversion of rimage to a submodule broke tarball releases. Interesting no one noticed - does anyone still use tarballs?

**Comment by lgirdwood** [comment]
@marc-hb probably not too many folks, but it does need fixed. Can you take a look and send a PR. Thanks

**Comment by marc-hb** [comment]
> > the conversion of rimage to a submodule broke tarball releases (cause they miss rimage entirely)

I did some research and found a couple ways to fix this but no simple and easy way. Long story short some sort of script that iterates on the submodule(s) is required. There is already a couple of "git-archive-all" scripts here and there already but nothing "standard" and certainly nothing built in git; so this would be a new/extra dependency and not necessarily a well maintained one.

The  other packaging option I considered is CMake's CPack. It's standard in CMake and should do the job, however it's wildcard-based and not version control based and requires a pristine checkout. Mmmm....

BTW I got confirmation on Slack from @mbolivar himself (thx) that `west` (used by Zephyr instead of submodules) doesn't support source releases more than git submodules do. In other words, any west project that wants to create a source tarball has to iterate on the list of git clones too. This feature has never been requested in west - another demonstration very few people still care about release tarballs?

**In conclusion I think doing nothing and waiting** for someone to ask is probably a good option because:
- source tarballs seem to have few users left
- those few tarball users left can perfectly treat rimage as a standalone project and external dependency, which is more or less the point of a submodule. I made sure "extracting" the rimage source code inside `sof-v1.2.3/` Just Works. A symbolic link to wherever works too. Missing rimage produces a very good and prompt "rimage not found" error message (only the suggestion to use git submodules becomes off the mark)

We could even drop the symbolic link thanks to some new `cmake -DRIMAGE_LOCATION=~/rimage` CMake option if desired; should be easy enough.

**Comment by lgirdwood** [comment]
@marc-hb ack - I've no objection to dropping tarball support.

**Comment by marc-hb** [comment]
Thanks! To be shorter and hopefully clearer: as long as you manually re-add the `rimage/` directory yourself, the source tarball still works perfectly fine.

**Comment by marc-hb** [comment]
First submodule "usability" issue, I'm surprised it took that long: #3200

**Comment by plbossart** [comment]
well, I broke it, but it's not like I didn't warn folks that git submodules were the worst idea since ClearCase: https://github.com/thesofproject/sof/pull/2086#issuecomment-553018493

**Comment by marc-hb** [comment]
Never replace a directory with a submodule _of the same name_ because this creates a "flag day" that breaks git checkout, git bisect and upsets everyone including @plbossart. Demonstration:

```
cd sof/
git stash
git checkout bf795c5b249d
error: The following untracked working tree files would be overwritten by checkout:
	rimage/CMakeLists.txt
Please move or remove them before you switch branches.
Aborting
```

This would not happen if this rimage submodule had been created with a different name; for instance `rimage-sub` or `submodules/rimage`. Anything but a name already used in the past.

It's too late now but IMHO good to know with submodules in general which is why I'm posting this comment.


In related news I'm proposing to drop submodules in SOF and use West instead, thx for discussing the proposal at 
https://github.com/orgs/thesofproject/teams/sof-developers/discussions/46

**Comment by marc-hb** [comment]
Suggestion to more rimage back to sof.git:
- #8178
